fix huge number of watches in zk issue

kaijianding opened the pull request:
fix #6647 .
I guess the origin Announcer.java is better performance for segment announcement, so I just keep it.

I guess the root cause of this issue is that each child path has a watch on zookeeper server even we only add listener at the parent path. When the path is like /druid/prod/announces/$host, the number of watches is n * n (each server watches every host including itself). Plus the path for lookup, the number is n * n * 2.

I test it in my product environment, the number of watches is reduced as expected.

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+            transaction = transaction.delete().forPath(parent).and();
+          }
+          catch (Exception e) {
+            log.info(e, "Unable to delete parent[%s], boooo.", parent);
```
log.error maybe

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+          ((CuratorTransactionFinal) transaction).commit();
+        }
+        catch (Exception e) {
+          log.info(e, "Unable to commit transaction. Please feed the hamsters");
```
log.error maybe

leventov replied:
If a production Druid cluster has issues I don't think cluster operators will be happy to see this humour in the logs.

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+import java.util.concurrent.CopyOnWriteArrayList;
+
+/**
+ * Announces single node on Zookeeper and only watch this node, which is different with Announcer
```
Announcer should be a javadoc link

leventov replied:
"which is different with Announcer" the idea is not complete. Please elaborate.

kaijianding replied:
`which is different with Announcer` means `NodeAnnouncer announces single node on Zookeeper and only watch this node`, while Announcer watches all child path, not only this node

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+  private final List<Announceable> toUpdate = new ArrayList<>();
+  private final ConcurrentMap<String, NodeCache> listeners = new ConcurrentHashMap<>();
+  private final ConcurrentMap<String, byte[]> announcements = new ConcurrentHashMap<>();
+  private final List<String> parentsIBuilt = new CopyOnWriteArrayList<String>();
```
The name and the purpose of this field is not clear

leventov replied:
This list is iterated only once. There is no win in using `CopyOnWriteArrayList` rather than a simple `ArrayList` protected by a lock.

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+  private final CuratorFramework curator;
+
+  private final List<Announceable> toAnnounce = new ArrayList<>();
+  private final List<Announceable> toUpdate = new ArrayList<>();
```
How is this different from `toAnnounce`?

kaijianding replied:
`toUpdate` is added when update() called before zk is connected, the path to update may or may not be added to `toAnnounce` before.
Most part of NodeAnnouncer's code logic is following the original Announcer.java

leventov replied:
Please annotate `@GuadedBy("toAnnounce")`.

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+   *
+   * @param path                  The path to announce at
+   * @param bytes                 The payload to announce
+   * @param removeParentIfCreated remove parent of "path" if we had created that parent
```
The explanation of this parameter is not clear.

kaijianding replied:
this part of code is following the original Announcer's code.
this means only who created the parent path should do the remove. According to the git log, this code is added to fix a bug.

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+
+    final ZKPaths.PathAndNode pathAndNode = ZKPaths.getPathAndNode(path);
+
+    final String parentPath = pathAndNode.getPath();
```
Please extract it to a static utility method `ZKPathsUtils.getParentPath(path) { return ZKPaths.getPathAndNode(path).getPath(); }`, because currently the code looks confusing.

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+          buildParentPath = true;
+        }
+      }
+      catch (Exception e) {
```
Is it possible to catch a narrower exception type?

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+        }
+      }
+      catch (Exception e) {
+        log.debug(e, "Problem checking if the parent existed, ignoring.");
```
Why ignoring?

kaijianding replied:
this part of code is following the original Announcer's code.

when this exception happens: 1, if the parent path exists, this exception can be ignored. 2 if the parent path doesn't exist, the following code will throw exception due to failed to add path

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+                  if (currentData == null) {
+                    final byte[] value = announcements.get(path);
+                    if (value != null) {
+                      log.info("Node[%s] dropped, reinstating.", path);
```
Why it means that the node is dropped?

kaijianding replied:
like someone dropped it from zookeeper cli

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+        createAnnouncement(path, bytes);
+      }
+      catch (Exception e) {
+        throw Throwables.propagate(e);
```
Throwables.propagate() should not be used in new code.

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+    }
+
+    boolean created = false;
+    synchronized (toAnnounce) {
```
Several distinct synchronized blocks on the same variable in the same method doesn't seem to make a lot of sense to me

leventov replied:
@kaijianding could you please explain?

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+        }
+      }
+      catch (Exception e) {
+        throw Throwables.propagate(e);
```
Throwables.propagate() should not be used in new code.

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+      try {
+        createAnnouncement(path, bytes);
+      }
+      catch (Exception e) {
```
Is it possible to narrow the exception type?

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+          updateAnnouncement(path, bytes);
+        }
+      }
+      catch (Exception e) {
```
Is it possible to narrow the exception type?

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+      log.debug("Created parentPath[%s], %s remove on stop.", parentPath, removeParentsIfCreated ? "will" : "will not");
+    }
+    catch (Exception e) {
+      log.info(e, "Problem creating parentPath[%s], someone else created it first?", parentPath);
```
log.error

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+      if (removeParentsIfCreated) {
+        parentsIBuilt.add(parentPath);
+      }
+      log.debug("Created parentPath[%s], %s remove on stop.", parentPath, removeParentsIfCreated ? "will" : "will not");
```
What does "stop" mean here?

kaijianding replied:
stop is the stop() method annotated with LifecycleStop

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+    catch (KeeperException.NoNodeException e) {
+      log.info("node[%s] didn't exist anyway...", path);
+    }
+    catch (Exception e) {
```
Same as above

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+      log.info("node[%s] didn't exist anyway...", path);
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
```
Same as above

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+    try {
+      cache.start();
+    }
+    catch (Exception e) {
```
Same as above

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,346 @@
+    }
+    catch (Exception e) {
+      CloseQuietly.close(cache);
+      throw Throwables.propagate(e);
```
Same as above

kaijianding commented:
@gianm I'm not familiar with curator code, but it seems because we watch the parent path to get child events, curator adds watch for each child path

jihoonson commented:
Hi @kaijianding, I still don't understand why you have such many watchers. In Druid, only master nodes (brokers, coordinator, overlord if you use zk-based remoteTaskRunner) watch such announcements. For example, `AbstractCuratorServerInventoryView` watches `/some/zk/path/announcements` which is used only in brokers and the coordinator. Similarly, `/some/zk/path/listeners/lookups/` is watched only by the coordinator. Are you seeing that other modules like tasks are also watching those zk paths?

kaijianding commented:
@jihoonson Announcer.java is also used in realtime node to re-announce itself when temporary disconnecting to zk, Then a realtime node also watches `/some/zk/path/announcements`

leventov commented:
@kaijianding most answers to my questions that you given should be transformed into code comments, or the code should be altered in some other way. The objective is to approach the state that if I read the code for the first time, none of my questions were arisen.

jihoonson commented:
@kaijianding thanks. Would you please fix the conflicts and check @leventov's last comment?

kaijianding commented:
@jihoonson @leventov @gianm  conflicts are resolved and code comments are updated, do you have any more comment?

leventov commented:
@kaijianding this comment: https://github.com/apache/incubator-druid/pull/6683#issuecomment-453068505 is mostly not addressed.

Please don't force push: https://github.com/apache/incubator-druid/blob/master/CONTRIBUTING.md#if-your-pull-request-shows-conflicts-with-master

kaijianding commented:
@leventov I added comments to toAnnounce, toUpdate, parentsIBuilt, and modified some other comments to help better understanding. Could you tell me which part of comments you think can improve?

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,349 @@
+ * NodeAnnouncer announces single node on Zookeeper and only watches this node,
+ * while {@link Announcer} watches all child paths, not only this node
+ */
+public class NodeAnnouncer
```
Could you please add concurrent control flow documentation to this class? From what threads (thread pools, executors) each method may be called?

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,349 @@
+
+  // incase a path is added to `toAnnounce` in announce() before zk is connected,
+  // should remember the path and do announce in start() later
+  private final List<Announceable> toAnnounce = new ArrayList<>();
```
Please annotate `@GuadedBy("toAnnounce")`. Could use create a separate lock object in a field called "lock"?

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,349 @@
+    synchronized (toAnnounce) {
+      try {
+        if (!Arrays.equals(oldBytes, bytes)) {
+          announcedPaths.put(path, bytes);
```
Looks like a race condition is possible between `announcedPaths.get()` and `announcedPaths.put()`, is that benign? If not, please rewrite the code using `map.compute()`.

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,349 @@
+
+    boolean created = false;
+    synchronized (toAnnounce) {
+      if (started) {
```
It seems that `started` flag is already checked at the beginning of this method, why it is checked again here?

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,349 @@
+              }
+          );
+
+          if (started) {
```
It seems that `started` flag is already checked at the beginning of this method, why it is checked again here?

leventov commented on server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java:
```diff
@@ -0,0 +1,349 @@
+      }
+      CloseQuietly.close(closer);
+
+      for (String announcementPath : announcedPaths.keySet()) {
```
Entries are removed via `map.remove()` while iterating over the map. I cannot find proof that this is safe; it would be better to remove them via `iterator.remove()`.

leventov commented:
@kaijianding do you plan to continue this work?

jihoonson commented:
@kaijianding would you please fix the conflicts and check @leventov's latest comments? BTW, I'm untagging milestone since this issue is not necessarily a release blocker. Feel free to let me know if you think this should be.

pzhdfy commented:
![image](https://user-images.githubusercontent.com/2594619/72202190-8f790300-3497-11ea-8c6d-e53a5198e5fc.png)
We use this patch in our production environment，the watch count  significantly decreased .
